// alexa-req/src/index.js
export default class AlexaRequest {
  constructor({ session, request }) {
    this.session = session;
    this.request = request;
  }

  // returns the sessionId
  getSessionId() {
    return this.session.sessionId;
  }

  // returns the userId
  getUserId() {
    return this.session.user.userId;
  }

  // returns the user access token
  getUserAccessToken() {
    return this.session.user.accessToken;
  }

  // returns the user access token
  getTimeStamp() {
    return new Date((this.request.timestamp || '').replace(/-/g, '/').replace(/[TZ]/g, ' '));
  }

  // returns the request type
  getRequestType() {
    return this.request.type;
  }

  // returns the name of the intent
  getIntentName() {
    if (this.getRequestType() === 'IntentRequest') {
      return this.request.intent.name;
    }

    return this.getRequestType();
  }

  // will return the value of a slot if it exists
  getSlotValue(slot) {
    const aSlot = this.request.intent.slots[slot];
    if (!aSlot) {
      throw new Error(`Invalid slot ${slot}. Unable to find the slot in the request.`);
    }

    return aSlot.value;
  }

  // validates that the request has the correct applicationId
  verifyAppId(id) {
    return this.session.application.applicationId === id;
  }

}
